### ISSUE TITLE
Provable Interactive Transaction

### ISSUE THREAD

**Issue by garyyu** [mimblewimble/grin#2652]
# 1. Use case

## 1.1 In Bitcoin

In the Bitcoin transaction scenario, Bob tell Alice his Bitcoin address, for example: [3ChVP627...](https://www.blockchain.com/btc/address/3ChVP627KU5w4zu2rieFPF3wGXWQgmhvrs), and Alice do a Bitcoin transaction to pay to this address, for example: [3169d1b5...](https://www.blockchain.com/btc/tx/3169d1b575955648694da900d3f8d6aed34e65cdcf79f3ed16e6996788fda714), and the ***HASH160*** form of this address is directly put on the public chain database (as the Output Scripts): ```HASH160 PUSHDATA(20)[78c04dda49b06260020a106f8e279b1bfbadcb71] EQUAL```. 

This is a perfect **provable** transaction, since there's no way for Bob to deny he received this payment.

## 1.2 In Grin

In the Grin transaction scenario, Bob tell Alice his Grin wallet address, for example: `https://bob.grinwallet.me:3415`, and Alice do a Grin transaction to pay to Bob's said wallet. Once the transaction is done and confirmed on the chain, Bob has his UTXO on the chain, for example: [0861105a...](https://grinscan.net/output/0861105a99353111dc62f01708ed5b31d68ec6d6a9cebc2a46491821525883e4e3).

But the practical problem is: there's no any public indicator on the chain which can prove this output belongs to Bob. ONLY Bob can prove this output belongs to himself, nobody else without Bob's private key!

So, in case Bob **deny** he received Alice's payment, it will be very complex for Alice to prove she paid, and even impossible if she can't force Bob to provide related evidence.

# 2. Provable Interactive Transaction solution

## 2.1 Recap of Grin current slate structure in the Interactive Transaction

Let's take a look at an example of a finalized slate (of a wallet transaction):
```Javascript
{
  "num_participants": 2,
  "id": "8ba033b6-3e05-4cad-81ec-937f3056b1e8",
  "tx": {
    "offset": [...],
    "body": {
      "inputs": [
        {
          "features": "Plain",
          "commit": [...],
        }
      ],
      "outputs": [
        {
          "features": "Plain",
          "commit": [...],
          "proof": [...],
        },
        {
          "features": "Plain",
          "commit": [...],
          "proof": [...],
       }
      ],
      "kernels": [
        {
          "features": "HeightLocked",
          "fee": 8000000,
          "lock_height": 18187,
          "excess": [...],
          "excess_sig": [...],
       }
      ]
    }
  },
  "amount": 110000000,
  "fee": 8000000,
  "height": 18187,
  "lock_height": 18187,
  "participant_data": [
    {
      "id": 0,
      "public_blind_excess": [...],
      "public_nonce": [...],
      "part_sig": [...],
      "message": null,
      "message_sig": null
    },
    {
      "id": 1,
      "public_blind_excess": [...],
      "public_nonce": [...],
      "part_sig": [...],
      "message": null,
      "message_sig": null
     }
  ]
}
```
For the detail meaning of this structure, please refer to https://github.com/mimblewimble/grin/blob/391e311f4c05619bd4b7b0d599a5071ce7753b90/wallet/src/libwallet/slate.rs#L104:L127

## 2.2 Enhance Grin transaction as a provable transaction

The basic idea is to import the exact same **"address"** concept as Bitcoin, but only use it on the Grin transaction, not on the chain.

Still with above example slate, but with the enhanced fields:
```diff
{
  "num_participants": 2,
  "id": "8ba033b6-3e05-4cad-81ec-937f3056b1e8",
  "tx": {
    "offset": [...],
    "body": {
    (exact same as the original format)
    }
  },
  "amount": 110000000,
  "fee": 8000000,
  "height": 18187,
  "lock_height": 18187,
  "participant_data": [
    {
      "id": 0,
      "public_blind_excess": [...],
      "public_nonce": [...],
      "part_sig": [...],
      "message": null,
      "message_sig": null,
+     "p2pkh_addr": null,
+     "wallet_url": null,
+     "p2pkh_public_key": null,
+     "p2pkh_msg": [...],
+     "p2pkh_sig": [...]
    },
    {
      "id": 1,
      "public_blind_excess": [...],
      "public_nonce": [...],
      "part_sig": [...],
      "message": null,
      "message_sig": null,
+      "p2pkh_addr": [...],
+      "wallet_url": [...],
+      "p2pkh_public_key": [...],
+      "p2pkh_msg": [...],
+      "p2pkh_sig": [...]
     }
  ]
}
```

Here is the proposal for this provable Grin transaction procedure.
### 2.2.1 Wallet URL with the receiver's "address"

For example, when Bob tell Alice his Grin wallet address, he give both his wallet URL and his "address": `https://bob.grinwallet.me:3415?addr=3ChVP627KU5w4zu2rieFPF3wGXWQgmhvrs`.

This "address" has exact same format as Bitcoin's legacy address.

And an optional format for a Grin receiver's address is the URL (`https://bob.grinwallet.me:3415` in this example) and a separated "address" (`3ChVP627...` in this example) as a `wallet send` command parameter.

### 2.2.2 Step 1: Alice send Bob a slate

```diff
{
  "num_participants": 2,
  "id": "8ba033b6-3e05-4cad-81ec-937f3056b1e8",
  "tx": {
    "offset": [...],
    "body": {
    (exact same as the original format and content)
    }
  },
  "amount": 110000000,
  "fee": 8000000,
  "height": 18187,
  "lock_height": 18187,
  "participant_data": [
    {
      "id": 0,
      "public_blind_excess": [...],
      "public_nonce": [...],
      "part_sig":  null,
      "message": null,
      "message_sig": null,
+     "p2pkh_addr": null,
+     "wallet_url": null,
+     "p2pkh_public_key": null,
+     "p2pkh_msg": [...],
+     "p2pkh_sig": [...]
    }
  ]
}
```
As the sender / payer, Alice:
- leave `p2pkh_public_key`, `wallet_url` and `p2pkh_addr` as `null`
- put the `p2pkh_msg` as the HASH of this slate data (exclude  `p2pkh_*` and all other `participant_data` array elements, and exclude `id`)
- and put `p2pkh_sig` as the signature of this `p2pkh_msg` with the private key related to `public_blind_excess`. 

**Note**: Alice must keep the told Bob's `wallet_url` and `p2pkh_addr` as somewhere, and need to use them in the 3rd step.

### 2.2.3 Step 2: Bob receive above slate and send back the processed slate to Alice

**Note**: for security, Bob need check the received payment request whether his own "address" is there. If no, giving warning at this moment for compatibility of old version, and in the future we close the connection or ban the sender's IP in this case.

```diff
{
  "num_participants": 2,
  "id": "8ba033b6-3e05-4cad-81ec-937f3056b1e8",
  "tx": {
    "offset": [...],
    "body": {
    (exact same as the original format and content)
    }
  },
  "amount": 110000000,
  "fee": 8000000,
  "height": 18187,
  "lock_height": 18187,
  "participant_data": [
    {
      "id": 0,
      (same data as above received slate)
    },
    {
      "id": 1,
      "public_blind_excess": [...],
      "public_nonce": [...],
      "part_sig": [...],
      "message": null,
      "message_sig": null,
+      "p2pkh_addr": [...],
+      "wallet_url": [...],
+      "p2pkh_public_key": [...],
+      "p2pkh_msg": [...],
+      "p2pkh_sig": [...]
     }
  ]
}
```
As the receiver / payee, Bob: 
- **must** put the correct `wallet_url` and `p2pkh_addr` which must be exact same as what he told Alice for this payment
- put `p2pkh_public_key` as the one which match the `p2pkh_addr`
- put the `p2pkh_msg` as the HASH of this slate data (exclude  `p2pkh_*` and all other `participant_data` array elements, and exclude `id`)
- and put `p2pkh_sig` as the signature of this `p2pkh_msg` with the private key related to `p2pkh_public_key`. 

 ### 2.2.4 Step 3: Alice receive above slate and generate a finalized slate

```diff
{
  "num_participants": 2,
  "id": "8ba033b6-3e05-4cad-81ec-937f3056b1e8",
  "tx": {
    "offset": [...],
    "body": {
    (exact same as the original format and content)
    }
  },
  "amount": 110000000,
  "fee": 8000000,
  "height": 18187,
  "lock_height": 18187,
  "participant_data": [
    {
      "id": 0,
      "public_blind_excess": [...],
      "public_nonce": [...],
+     "part_sig": [...],
      "message": null,
      "message_sig": null,
      "p2pkh_addr": null,
      "wallet_url": null,
      "p2pkh_public_key": null,
+     "p2pkh_msg": [...],
+     "p2pkh_sig": [...]
    },
    {
      "id": 1,
      (same data as above received slate)
     }
  ]
}
```

As the sender / payer, Alice: 
- **must** check the above received slate for the `wallet_url` and `p2pkh_addr`, which must be exact same as what she was told by Bob for this payment
- **must** check the `p2pkh_addr` is indeed generated by `p2pkh_public_key`
- create her own partial signature for this transaction and update the `tx body`
- update her `p2pkh_msg` and `p2pkh_sig`
- keep this finalized slate stored at her wallet database, for any possible "provable" request in the future.

### 2.2.5 Finish

Alice post this transaction to the chain and this transaction will be confirmed at a few minutes.

## Alice prove she did this transaction

For example, Bob deny and say he never receive this transaction from Alice.

Alice show / publish her saved finalized slate for this transaction, to Bob or any other 3rd party, anyone can validate this data and confirm Bob's lying, since this slate include Bob's address and Bob's signature, and the integrity checking.

# Grin "address" generation

Since we already have BIP-32 HD wallet feature, it should be an easy point to have this / these "address"/es.

Btw, to support the new "address"/es, Grin wallet need to store all published / used "address", to enable the checking of any receiving payment requests, either on current "address" or on any previous "address".

# Other use cases

I'm thinking it seems possible to share the wallet database with the auditor/s, without telling the wallet seed / private key. The auditor/s still can get the logs which can be easily validated. To be discussed with a separated issue / topic.

# Validation API

We need some new API to enable an easy validation on this "provable" slate.

**Welcome your review and any comments :-)**
**Especially for the security level with this solution. Please feel free to correct me if any mistakes**

**Comment by jaspervdm** [comment]
In general I am definitely in favor of some kind of PKI to be added to slates. However I do have some comments:
- The nice thing about the current slate system is that it is completely agnostic of the transfer method. It can be HTTPS, file, grinbox, whatever. However, adding a `wallet_url` would remove that feature. We should try and figure out a way to keep it completely agnostic
- Instead of having `p2pkh_msg` and `p2pkh_sig`, couldn't we just sign the kernel excess? The presence of the kernel in the kernel set, coupled with a signature on it for the pubkey would prove that the payment indeed happened
- The grin `send` command could take an extra flag, `-a` that is the expected payment address (pubkey) and should be encoded in the slate somewhere as `expected_pubkey`. When the slate is filled by the other party, it should sign with the `expected_pubkey`. If they didn't, the slate should be denied

**Comment by garyyu** [comment]
@jaspervdm thanks for your review :-)

1. Agree we need figure out a way to keep is agnostic 👍 
2. Regarding the use of `the presence of the kernel in the kernel set`, I'm afraid I strongly do not suggest this direction. Because I'm still dreaming that there will be a day that we can find a nice way to throw the old kernels, just like throwing the spent outputs 😄 
3. Regarding the `extra flag`, I hope the provable transaction becomes the default behavior in the future, so perhaps a mandatory expected payment address is better?

**Comment by jaspervdm** [comment]
> 2. Regarding the use of `the presence of the kernel in the kernel set`, I'm afraid I strongly do not suggest this direction. Because I'm still dreaming that there will be a day that we can find a nice way to throw the old kernels, just like throwing the spent outputs 😄

So in your proposal you sign the whole slate (inputs, outputs, kernels). How does one validate this proof if all this information is no longer on-chain?

> 3. Regarding the `extra flag`, I hope the provable transaction becomes the default behavior in the future, so perhaps a mandatory expected payment address is better?

Yeah, we could combine the `pubkey` and `destination` fields, maybe something like `https://pubkey@hostname:port`, `keybase://pubkey@username` etc

**Comment by garyyu** [comment]
> So in your proposal you sign the whole slate (inputs, outputs, kernels). 

Sign the whole slate (more accurately, not real whole but indeed including the whole tx body: inputs, outputs, kernels) is also for the integrity checking of this whole slate.

By signing the whole slate, when the sender show / publish his/her saved finalized slate for a transaction, anybody can easily verify whether it is a true whole slate without any modification.

Also, we need check the aggregated Schnorr signature (from the partial sigs) indeed get the final excess_sig in the kernel.

> How does one validate this proof if all this information is no longer on-chain?

No, the (inputs, outputs) could be no longer on-chain. But the kernel definitely can be found on archive nodes which keep all past blocks data for ever, even when we start throwing old kernels in the future just like the spent outputs. And on lower security level, we can use multiple GRIN explorers to check kernel existence.

So, to summarize, the idea for validating this proof is:
- Slate integrity checking (by hashing to get the signing message)
- Receiver's address signature checking (p2pkh_addr, p2pkh_public_key, p2pkh_msg, p2pkh_sig)
- Schnorr signature aggregation checking ( partial sigs, excess_sig)
- kernel signature checking
- kernel excess existence (on chain) checking

Also, the sender prove the `wallet_url` is indeed the one which the receiver told him/her.


> Yeah, we could combine the `pubkey` and `destination` fields, maybe something like `https://pubkey@hostname:port`, `keybase://pubkey@username` etc

👍  will modify the proposal to take this :-)

**Comment by Agreene** [comment]
I like the idea!
One issue I can think of is that Bob could later simply deny that he is the owner of the wallet bob.grinwallet . Or alternatively, Alice could create a wallet fakebob.grinwallet, complete the transaction with herself, and then claim that she sent it to Bob.

For this to work the parties involved would need to only use public declared, undeniable wallet addresses.  Which is a difficult task if it's not stored on the chain.

**Comment by garyyu** [comment]
@Agreene 
> One issue I can think of is that Bob could later simply deny that he is the owner of the wallet bob.grinwallet

That's a common "problem" in life. Same in Bitcoin, Bob can deny that he is the owner of the Bitcoin address which he just told Alice before. This Provable Interactive Tx solution is not for that.

And I think no blockchain solution can solve the problem you mentioned above. But a simple paper contract can solve it :-)   just write down the Bob's Bitcoin address on the contract, then both Bob and Alice write a handwriting signature on the contract 😄 but surely it's not related to this "Provable" Tx discussion.

**Comment by GandalfThePink** [comment]
I think that this is a great idea. 

In addition to the signed slate is should also be checked that the slate has actually made it into the chain. Or am I missing any reason why the signed slate on its own is sufficient. 

Proving transaction existence on the chain could be difficult when the UTXO in question is already spent again, potentially requiring Merkle proofs.  

Please also have a look at https://github.com/mimblewimble/grin/issues/2631 where I integrate the 'address' off the receiver directly into the transaction and thereby in the chain. This is however suboptimal as it potentially leaks the amounts sent. But maybe there still is some merit to this?

**Comment by garyyu** [comment]
@GandalfThePink thanks for your review.

> the signed slate is should also be checked that the slate has actually made it into the chain. Or am I missing any reason why the signed slate on its own is sufficient.

As said in https://github.com/mimblewimble/grin/issues/2652#issuecomment-469948016, `kernel excess existence (on chain) checking` is mandatory, make sure the kernel part of this signed slate is indeed on the chain.

> Proving transaction existence on the chain could be difficult when the UTXO in question is already spent again, potentially requiring Merkle proofs.

Yes, that's why we need check the kernel instead of the output.

> Please also have a look at #2631 where I integrate the 'address' off the receiver directly into the transaction and thereby in the chain.

Okay, will look into #2631 and could leave comments there.

> This is however suboptimal as it potentially leaks the amounts sent. But maybe there still is some merit to this?

You can't say it `leaks` the amounts sent 😄   To prove a transaction, it's the purpose of this signed slate.

And the payer always keep this signed slate, but it doesn't mean he/she always need to publish / send this to anyone.  This is for the exception case when **the receiver denied** 😄  which should be very low ratio but indeed very very important!

**Comment by GandalfThePink** [comment]
@garyyu 
In fact I think that the idea you propose is superior to mine, because it better protects the amounts from external parties. I also missed that in this case just the kernel is sufficient for verification which makes this much easier than tracking UTXO's or when they are spent Merkle proofs thereof. 

Therefore I think that my older issue can be closed and we should move ahead with this idea.

**Comment by yeastplume** [comment]
I don't think this belongs at the core transaction layer, nor do I think it's desirable to allow a recipient to collect irrefutable proof of who sent them what without the sender's explicit consent.

As far as the proposed changes, the basic problem here is that this solution is trying to come up with a homebrew method of proving that somebody is who they say they are. This is a very complex task, and there is a multitude of solutions that already attempt do this. 

If I'm a merchant and I want to prove who I am, I likely already have something in place. I have some signing cert that verifies up to chain to the usual root certs, and the public key is prominent on my website If I'm interested in proving a transaction was definitely me I can sign a combination of the kernel offset and amount with that cert, put it in the existing message field, and anyone can verify using existing infrastructure. Similarly, if I'm a keybase user or have any other system in place that I like to use for verifying myself, I can do the same thing. My point here is the concept of 'address' should not be enforced or encoded into the transaction exchange protocol level.

Also, if I send Bob 10 Grins and I later want to deny that I did so, that is absolutely Bob's right. Imagine a Ashley Madison takes payment in Grin, then its db gets hacked and all of a sudden the attacker has a handly list of addresses along with irrefutable proof where the payment came from.

If all parties in a transaction want to prove who they are and trust the others to keep the information appropriately, then by all means they should be allowed to do so, and indeed there already are ways they can. I'm okay with providing lightweight and non-prescriptive helpers to facilitate this, (like provide a particular message with a combination of fields to sign to prove ownership and amounts, as well as a verify function,) but it's going to take a lot of convincing to make me think we should be embedding addresses in any format directly in the transaction by default.

**Comment by GandalfThePink** [comment]
I think what we aim to implement is that when Alice pays Bob, she can later prove that to Carol, where Carol was not involved in transaction building. 

To do so, Bob needs to identify and this can be done via a public key. Bob actively signs the slate and could refuse to do so when he does not want to be linked in any way to the payment. Therefore Bob voluntarily opts in to having this transaction verifiable. 

Alice can now prove to Carol that she has payed an entity with possession of the private key to Bobs public key. This proof is initiated by Alice and she could still choose to not verify and have plausible deniability. 
Carol by herself does not see anything special on the chain and the transactions in question looks indistinguishable to any other transaction before receiving proof from Alice. 

Therefore I do not see any privacy flaws in this implementation. Furthermore it does only modify the wallet functionality and leave the base chain level unaffected. I think it does make sense to implement such a feature at the wallet level since Bob has to commit to the transaction before it is finalised. It is not sufficient to let Bob sign something afterwards since Bob could simply refuse to do so and then deny payment from Alice. The implementation needs to be 'atomic' and this is best achieved at the transaction building level. 
After-all the goal is to give Alice the power to prove a transaction independent of Bob.

**Comment by garyyu** [comment]
@yeastplume 

> Also, if I send Bob 10 Grins and I later want to deny that I did so, that is absolutely my right. 

Absolutely 😄  
But it's NOT the case we're discussing on this "Provable" Tx solution!
In this solution, we're trying to make **Bob** / **Receiver** undeniable, NOT Alice.

> Imagine an Ashley Madison takes payment in Grin, then its db gets hacked and all of a sudden the attacker has a handy list of addresses along with irrefutable proof where each payment came from.

Same as above! this Ashley Madison is **receiver**, Ashley Madison can't deny received user's payment, but the user / payer  CAN deny, with this "Provable" Tx solution.

**Comment by JacobPlaster** [comment]
+1 This would be another great feature for us!

**Comment by ignopeverell** [comment]
To start with, I think we can state a few things to simplify the setup a bit, with Alice sending to Bob and Carol being some sort of arbitrator or external observer:

1. Bob can always deny being involved in the transaction, this is generally true in real life as well as in other cryptocurrencies, as long as some sort of public cert isn't involved (as @yeastplume mentioned). So all we can help with is for Alice to show to Carol that she did send a given amount at a given time. Then it's up to Carol to check the amount match those Alpaca socks Alice never received.
2. I don't think we should worry too much about disappearance from the chain. Either the dispute is going to be done in a fairly timely fashion or archival nodes can get involved. Worst case we can do Merkle paths if required but I don't think it's necessary.

So with that in mind, what's wrong with Alice providing:

1. The full transaction (saved in her wallet).
2. Amounts for all inputs and her change output.
3. Pubkeys and signatures to the empty string for all our inputs and her change output.

That should be enough to convince Carol without any change or complication. Or am I missing something?

**Comment by GandalfThePink** [comment]
@ignopeverell In this case Alice could simply pay the amount to herself and claim that this was the payment for the Alpaca socks. 

Instead when the merchant of the socks provides a public key and signs the slate with the corresponding private key, Alice can prove that she has actually paid the merchant (and not only made a fake payment). This is proof that the owner of said public key was involved in the transaction and Bob can no longer deny that. 

When we have merchants operating on Grin, I think this is a feature that most customers will want to have. After all they are taking the risk and to mitigate that risk they should have proof of the Merchant having received the money.

**Comment by kargakis** [comment]
I am torn; not a big fan of PKI because to use it securely, vendors really need to be using one-time pairs, otherwise they risk being charged for products they never sold in the event their key gets leaked. I like the simplicity of doing away with this proposal in favor of reusing a serving cert but its leakage is even worse for vendors than a normal serving cert.

Also, vendors can still deny they didn't receive a transaction by claiming their key was leaked, right?

**Comment by GandalfThePink** [comment]
@kargakis Yes vendors can still claim their key was leaked. But that would be their own fault and they probably should take responsibility for this. Also just leaking the key is not enough. The potential attacker must also convince Alice to pay to them and not to the original vendor. If anything, the added signature improves the security for the vendor. 

I am thinking how Grin payments would fit into a legal framework (not necessarily a national one, also something like bisq or smart contracts). In the current case without any proof, all risk is sitting at the buyer. The buyer is committing money, then the merchant may deliver or not. And in case the merchant does not deliver no third party can objectively settle the conflict. But we can add functionality for proof of payment and in case the merchant has some mechanism for proof of delivery (which is not our problem) a third party can settle. This third party might be a smart contract, funded by the merchant that simply releases money to the buyer in case of conflict.

**Comment by kargakis** [comment]
@GandalfThePink I thought the main reason behind this proposal was to protect buyers against malicious vendors. From @garyyu's initial post:

> This is a perfect provable transaction, since there's no way for Bob to deny he received this payment.

It is not perfect and it is also questionable whether it can help in most cases a dispute will arise where a vendor denies they received a payment. If a vendor **really** wants to dispute a transaction, why wouldn't they go the extra step and claim their key was leaked? Any evidence a sender gives to an observer can be claimed to be fake at that point?

**Comment by GandalfThePink** [comment]
I would hope that a third party will refund the buyer even when the vendor claims that keys were leaked. Not leaking keys is the responsibility of the vendor. But that is not our job to decide.

I think that the current Grin architecture allows these proofs and that there are use-cases for them. (See for example bisq). Over the long run they will get built.

**Comment by ignopeverell** [comment]
Here we're all assuming some sort of pubkey that's been widely share, similarly to a SSL cert or address. In which case I have to agree with @yeastplume it doesn't seem to fit in the core wallet exchange protocol and maybe should be specified separately, by relying on a structured use of the comment field, for merchant wallets. AFAIK, most merchants already need specific wallet infrastructures for their specific needs.

So I'd suggest moving forward by:

1. If the comment field isn't enough, perhaps specify an opaque extension field for those additional use cases.
2. Specify a payment protocol for merchant wallets, so they can be interoperable (with a corresponding client protocol that wallets can choose to support).

**Comment by GandalfThePink** [comment]
I agree, specifying a protocol is a good idea and merchants can then choose to implement that. 

But also on the clients side there should be a check signature before the finalised transaction is broadcasted. The transaction could be build completely and we add this verification before publishing the transaction to the chain. In addition the wallet should also store the proof. Do you think it is reasonable to integrate this functionality into the base wallet, or should that be left for third-party wallets?

**Comment by kargakis** [comment]
[BIP70](https://github.com/bitcoin/bips/blob/master/bip-0070.mediawiki) is related to this discussion.
